source/MRMesh/MRBestFitPolynomial.cpp

```diff
@@ -0,0 +1,449 @@
+#include "MRBestFitPolynomial.h"
+#include "MRGTest.h"
+
+#include <Eigen/Dense>
+
+#include <cmath>
+#include <complex>
+
+namespace
+{
+
+
+template <typename T, size_t degree>
+struct Solver
+{
+    Eigen::Vector<std::complex<T>, degree> operator() ( const Eigen::Vector<T, degree + 1>& coeffs ) = delete;
+};
+
+template <typename T>
+struct Solver<T, 1>
+{
+    Eigen::Vector<std::complex<T>, 1> operator() ( const Eigen::Vector<T, 2>& c )
+    {
+        assert( c[1] != 0 );
+        // y(x) = c[0] + c[1] * x = 0 => x = -c[0] / c[1]
+        return Eigen::Vector<std::complex<T>, 1>{ -c[0] / c[1] };
+    }
+};
+
+template <typename T>
+struct Solver<T, 2>
+{
+    Eigen::Vector<std::complex<T>, 2> operator() ( const Eigen::Vector<T, 3>& coeffs )
+    {
+        assert( coeffs[2] != 0 );
+
+        // y(x) = c[0] + c[1] * x + c[2] * x^2 = 0
+        const auto b = coeffs[1] / coeffs[2];
+        const auto c = coeffs[0] / coeffs[2];
+
+        const auto D = std::sqrt( std::complex<T>{ b * b - 4 * c } );
+        return { ( -b + D ) / T( 2 ), ( -b - D ) / T( 2 ) };
+    }
+};
+
+template <typename T>
+struct Solver<T, 3>
+{
+    Eigen::Vector<std::complex<T>, 3> operator() ( const Eigen::Vector<T, 4>& coeffs )
+    {
+        assert( coeffs[3] != 0 );
+        const auto& a = coeffs[3];
+        const auto& b = coeffs[2];
+        const auto& c = coeffs[1];
+        const auto& d = coeffs[0];
+
+        const T p = ( 3*a*c - b*b ) / ( 3*a*a );
+        const T q
+            = ( 2*b*b*b - 9*a*b*c + 27*a*a*d ) / ( 27 * a*a*a );
+        const T alpha = b / ( 3 * a );
+
+        const std::complex<T> D = q*q / T(4) + p*p*p / T(27);
+        const auto Ds = std::sqrt( D );
+
+        const std::complex<T> e1 = ( std::complex<T>( T(-1), sqrt( T(3) ) ) ) / T(2);
+        const std::complex<T> e2 = ( std::complex<T>( T(-1), - sqrt( T(3) ) ) ) / T(2);
+
+        const auto u1s = std::pow( -q / T(2) + Ds, 1 / T(3) );
+        const auto u2s = std::pow( -q / T(2) - Ds, 1 / T(3) );
```

Comment:
Better `std::cbrt` than `pow( , 1/3 )`

Reply:
It doesn't exist for complex numbers, but `Ds` might be non-real.